vnext/Microsoft.ReactNative/Views/TextViewManager.cpp

```diff
@@ -72,73 +70,102 @@ class TextShadowNode final : public ShadowNodeBase {
       m_firstChildNode = nullptr;
     }
 
-    Super::AddView(child, index);
-
-    if (auto run = static_cast<ShadowNodeBase &>(child).GetView().try_as<winrt::Run>()) {
-      if (m_backgroundColor) {
-        AddHighlighter(m_backgroundColor.value(), m_foregroundColor, run.Text().size());
-      }
-      m_prevCursorEnd += run.Text().size();
-    } else if (auto span = static_cast<ShadowNodeBase &>(child).GetView().try_as<winrt::Span>()) {
-      AddNestedTextHighlighter(
-          m_backgroundColor, m_foregroundColor, span, static_cast<VirtualTextShadowNode &>(child).m_highlightData);
+    if (addInline) {
+      Super::AddView(child, index);
     }
+
+    UpdateTextHighlighters();
   }
 
-  void AddNestedTextHighlighter(
-      const std::optional<winrt::Windows::UI::Color> &parentBackColor,
-      const std::optional<winrt::Windows::UI::Color> &parentForeColor,
-      winrt::Span &span,
-      VirtualTextShadowNode::HighlightData highData) {
-    if (!highData.backgroundColor && parentBackColor) {
-      highData.backgroundColor = parentBackColor;
-    }
+  void removeAllChildren() override {
+    m_firstChildNode = nullptr;
+    Super::removeAllChildren();
+    UpdateTextHighlighters();
+  }
 
-    if (!highData.foregroundColor && parentForeColor) {
-      highData.foregroundColor = parentForeColor;
+  void RemoveChildAt(int64_t indexToRemove) override {
+    if (indexToRemove == 0) {
+      m_firstChildNode = nullptr;
     }
+    Super::RemoveChildAt(indexToRemove);
+    UpdateTextHighlighters();
+  }
+
+  void UpdateTextHighlighters() {
```

Comment:
Add logic to update highlighters when root text backgroundColor and foregroundColor change